### PR TITLE
loading indication for syncing

### LINKED ISSUE #233: Loading Indication for Initial Sync Import
When loading up my entries for the very first time, I should be given an affordance of something happening within the app, e.g., a loading indication, if it takes more than two seconds to load and import the data into the app.

## Visual Design
**InVision**
https://mozilla.invisionapp.com/share/BQEUSO0S2#/290701009_Entry_List_-_Syncing

**Zeplin**
https://zpl.io/2Ejdl78

### PR DESCRIPTION
**Pull request by sashei**
resolves #233

### CODE REVIEW

**Comment by devinreams** [comment]
>loading and enabling working as enumerated

Nice! 👏

File: lockbox-ios/Common/Extensions/UIViewController+.swift [hunk 1]
```diff
@@ -22,35 +23,20 @@ protocol AlertControllerView {
                                 style: UIAlertControllerStyle)
 }
 
+protocol SpinnerAlertView {
+    func displaySpinner(_ dismiss: Driver<Void>, bag: DisposeBag)
+}
+
 extension UIViewController: StatusAlertView {
```
Comment (jhugman) [inline lockbox-ios/Common/Extensions/UIViewController+.swift]: 😕 

Putting this on `UIViewController` is a little over-general. To be sure, not wrong, just… I'll keep reading.

File: lockbox-ios/Common/Extensions/UIViewController+.swift [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 import UIKit
 import RxSwift
+import RxCocoa
 
 protocol StatusAlertView {
```
Comment (jhugman) [inline lockbox-ios/Common/Extensions/UIViewController+.swift]: Nit: naming. This shouldn't be called `View` if it isn't a `UIView`.

Suggest: `StatusAlertDisplay`? `InfiniteProgressDisplay`?

File: lockbox-ios/Presenter/ItemListPresenter.swift [hunk 3]
```diff
@@ -9,9 +9,11 @@ import RxDataSources
 import Storage
 import Shared
 
-protocol ItemListViewProtocol: class, AlertControllerView {
+protocol ItemListViewProtocol: class, AlertControllerView, SpinnerAlertView {
```
Comment (jhugman) [inline lockbox-ios/Presenter/ItemListPresenter.swift]: This makes sense to have `SpinnerAlertView` here, but only if `UIViewController` isn't extended by `SpinnerAlertView`.

File: lockbox-ios/Presenter/ItemListPresenter.swift [hunk 4]
```diff
@@ -124,37 +135,26 @@ class ItemListPresenter {
...
+                        self.view?.displayFilterCancelButton()
+                    }
+                })
 
         let listDriver = self.createItemListDriver(
                 loginListObservable: itemListObservable,
                 filterTextObservable: filterTextObservable,
-                itemSortObservable: itemSortObservable
+                itemSortObservable: itemSortObservable,
+                syncStateObservable: self.dataStore.syncState
```
Comment (jhugman) [inline lockbox-ios/Presenter/ItemListPresenter.swift]: Oh neat.

File: lockbox-ios/Presenter/ItemListPresenter.swift [hunk 5]
```diff
@@ -196,13 +224,46 @@ extension ItemListPresenter {
...
     }
 
+    private func placeholderComputingItemListObservable() -> Observable<[Login]> {
+        let hideSpinnerObservable = self.dataStore.syncState
+                .filter {
+                    $0 == SyncState.Synced
+                }
+                .map { _ in
+                    return ()
+                }
```
Comment (jhugman) [inline lockbox-ios/Presenter/ItemListPresenter.swift]: Wut. Comment needed?

File: lockbox-ios/Common/Extensions/UIViewController+.swift (hunk 1 above)
Comment (sashei) [reply lockbox-ios/Common/Extensions/UIViewController+.swift]: What if we want to throw up a spinner somewhere else? :p

**Comment by sashei** [comment]
@nickbrandt I've addressed the smaller changes (sorting option disabled, no separator line on the empty state view). @changecourse and I discussed filing a follow-up issue for the visual state of the search bar, there's not a trivial way to do that unfortunately :( 

I think I got the placeholder sticking where it's supposed to be, let me know how it looks to you.

**Comment by sashei** [comment]
@nickbrandt I've addressed the smaller changes (sorting option disabled, no separator line on the empty state view). @changecourse and I discussed filing a follow-up issue for the visual state of the search bar, there's not a trivial way to do that unfortunately :( 

I think I got the placeholder sticking where it's supposed to be, let me know how it looks to you.

File: lockbox-ios/Common/Extensions/UIViewController+.swift (hunk 1 above)
Comment (jhugman) [reply lockbox-ios/Common/Extensions/UIViewController+.swift]: Right, I'm just seeing the other `StatusAlertView` and `AlertControllerView` on `UIViewController+.swift`.

If this was the first time you'd done this, I'd advise on an approach which `viewController.present(_ view: StatusAlertView, animated: Bool = false)`, with a convenience constructor on `StatusAlertView`.

However, this is an established pattern, so carry on.

**Comment by nickbrandt** [comment]
@sashei This looks great! Thanks for fixing those items, and sounds good on the disabled state 👍

**Pull request merged**